File: 06. Backtracking/03. Palindromic Partitioning/main.java
import java.util.*;

public class main {
    
    // Function to check if a substring is a palindrome
    private static boolean isPalindrome(String s, int start, int end) {
        while (start < end) {
            if (s.charAt(start) != s.charAt(end)) return false;
            start++;
            end--;
        }
        return true;
    }

    // Backtracking function to generate partitions
    private static void backtrack(int start, String s, List<String> current, List<List<String>> result) {
        if (start == s.length()) {
            result.add(new ArrayList<>(current));
            return;
        }

        for (int end = start; end < s.length(); end++) {
            if (isPalindrome(s, start, end)) {
                current.add(s.substring(start, end + 1)); // Take substring
                backtrack(end + 1, s, current, result);
                current.remove(current.size() - 1); // Backtrack
            }
        }
    }

    // Function to generate all palindromic partitions
    public static List<List<String>> partition(String s) {
        List<List<String>> result = new ArrayList<>();
        backtrack(0, s, new ArrayList<>(), result);
        return result;
    }

    // Utility function to print partitions
    private static void printPartitions(List<List<String>> partitions) {
        for (List<String> partition : partitions) {
            System.out.println("  " + partition);
        }
    }

    public static void main(String[] args) {
        String s = "aab";
        System.out.println("Palindromic Partitions of \"" + s + "\":");
        printPartitions(partition(s));
    }
}

File: 06. Backtracking/03. Palindromic Partitioning/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Function to check if a given substring is a palindrome
bool isPalindrome(const string &s, int start, int end) {
    while (start < end) {
        if (s[start] != s[end]) return false;
        start++;
        end--;
    }
    return true;
}

// Backtracking function to find all palindromic partitions
void backtrack(int start, string &s, vector<string> &current, vector<vector<string>> &result) {
    if (start == s.size()) {
        result.push_back(current);
        return;
    }

    for (int end = start; end < s.size(); end++) {
        if (isPalindrome(s, start, end)) {
            current.push_back(s.substr(start, end - start + 1)); // Take the substring
            backtrack(end + 1, s, current, result);
            current.pop_back(); // Backtrack
        }
    }
}

// Function to generate all palindromic partitions
vector<vector<string>> partition(string s) {
    vector<vector<string>> result;
    vector<string> current;
    backtrack(0, s, current, result);
    return result;
}

// Utility function to print the partitions
void printPartitions(const vector<vector<string>> &partitions) {
    for (const auto &partition : partitions) {
        cout << "  [";
        for (int i = 0; i < partition.size(); i++) {
            cout << "\"" << partition[i] << "\"";
            if (i < partition.size() - 1) cout << ", ";
        }
        cout << "]\n";
    }
}

int main() {
    string s = "aab";
    
    cout << "Palindromic Partitions of \"" << s << "\":\n";
    vector<vector<string>> partitions = partition(s);
    printPartitions(partitions);

    return 0;
}
